### PR TITLE
Limit propagation of contextual values

### PR DESCRIPTION
**Pull request by xmo-odoo**
Task 2366614

While the propagation of contextual values *can* be convenient, it usually goes way out of scope, silently affecting things in unexpected ways.

This PR tries to clean up two big items:

* greatly limiting the propagation of contextual defaults (`default_{field}`) from `create`, these should not get out of the `create`, or be used *within* it beyond the toplevel record(s) being created.
* limiting the scope of `active_test`, it probably should not be inherited by the *result* of the `search`

### CODE REVIEW

**Comment by xmo-odoo** [comment]
@tde-banana-odoo the fix to test_mail preserves the current behaviour but that behaviour seems odd: when the `template_id` is *unset*, the composer's `attachment_ids` are only reset IFF `default_attachment_ids` is set in the context, that seems... not very sane?

Shouldn't we either have an emptying `attachment_ids` default, or have `onchange_template_id` explicitly pass in `default_attachment_ids` (as it already does with several other values)?

File: odoo/models.py [hunk 1]
```diff
@@ -3823,8 +3823,12 @@ def create(self, vals_list):
...
                     data['stored'][parent_name] = parent.id
 
-        # create records with stored fields
-        records = self._create(data_list)
+        # create records with stored fields, dropping `default_{field}` keys
+        # from the env for sub-field creations as well as return
+        records = self.with_context({
+            k: v for k, v in self.env.context.items()
+            if not k.startswith('default_')
+        })._create(data_list)
```
Comment (Yajo) [inline odoo/models.py:3831]: You can use: https://github.com/odoo/odoo/blob/fa576fc74e904ea428fd9cd24070ddbb64c553b4/odoo/tools/misc.py#L980-L982

Comment (xmo-odoo) [reply odoo/models.py:3831]: Nice suggestion, thanks, and I can even easily update it to take an exclusion predicate for filtering out `active_test`.

File: odoo/addons/test_new_api/tests/test_create.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
+from odoo.tests.common import TransactionCase
```
Comment (Julien00859) [inline odoo/addons/test_new_api/tests/test_create.py:1]:
```suggestion
from odoo import Command
from odoo.tests.common import TransactionCase
```

File: odoo/addons/test_new_api/tests/test_create.py [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+class TestNoDefaultPropagation(TransactionCase):
+    def test_sub_create(self):
+        """ `default_{name}` keys should not be propagated to the creation of
+        sub-records
+        """
+        Model = self.env['test_new_api.model_parent_m2o'].with_context(default_name="STAFF SGT. MAX FUCKING FIGHTMASTER")
+
+        rec = Model.create({
+            'child_ids': [(0, 0, {})]
+        })
```
Comment (Julien00859) [inline odoo/addons/test_new_api/tests/test_create.py:12]:
```suggestion
        rec = Model.create({
            'child_ids': [Command.create({})]
        })
```

File: odoo/addons/test_new_api/tests/test_create.py [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+        self.assertEqual(r2.name, False)
+
+    def test_sub_create_weird(self):
+        """ If the field with a default is a related then nothing should happen,
+        and the field should still get stripped.
+        """
+
+        Model = self.env['test_new_api.multi'].with_context(default_name="Dr. Steel")
+
+        rec = Model.create({'lines': [(0, 0, {})]})
```
Comment (Julien00859) [inline odoo/addons/test_new_api/tests/test_create.py:33]:
```suggestion
        rec = Model.create({'lines': [Command.create({})]})

**Review by Julien00859 — APPROVED** [review]

File: odoo/addons/test_new_api/tests/test_create.py (hunk 4 above)
Comment (tde-banana-odoo) [reply odoo/addons/test_new_api/tests/test_create.py:33]: No !

File: odoo/addons/test_new_api/tests/test_create.py (hunk 3 above)
Comment (tde-banana-odoo) [reply odoo/addons/test_new_api/tests/test_create.py:12]: Argl !

**Pull request closed without merging**